File: contract-ts/src/contract.ts
// Find all our documentation at https://docs.near.org
import { NearBindgen, near, call, view, AccountId, NearPromise, initialize, assert, PromiseIndex } from "near-sdk-js";
import { predecessorAccountId } from "near-sdk-js/lib/api";

class Bid {
  bidder: AccountId;
  bid: bigint;
}

const FIVE_TGAS = BigInt("5000000000000");
const NO_DEPOSIT = BigInt(0);
const NO_ARGS = JSON.stringify({});

@NearBindgen({ requireInit: true })
class AuctionContract {
  highest_bid: Bid = { bidder: '', bid: BigInt(1) };
  auction_end_time: bigint = BigInt(0);
  auctioneer: string = "";
  auction_was_claimed: boolean = false;
  ft_contract: AccountId = "";

  @initialize({ privateFunction: true })
  init({ end_time, auctioneer, ft_contract }: { end_time: bigint, auctioneer: string, ft_contract: AccountId }) {
    this.auction_end_time = end_time;
    this.highest_bid = { bidder: near.currentAccountId(), bid: BigInt(0) };
    this.auctioneer = auctioneer;
    this.ft_contract = ft_contract;
  }

  // @call({ payableFunction: true })
  // bid(): NearPromise {
  //   // Assert the auction is still ongoing
  //   assert(this.auction_end_time > near.blockTimestamp(), "Auction has ended");

  //   // Current bid
  //   const bid = near.attachedDeposit();
  //   const bidder = near.predecessorAccountId();

  //   // Last bid
  //   const { bidder: lastBidder, bid: lastBid } = this.highest_bid;

  //   // Check if the deposit is higher than the current bid
  //   assert(bid > lastBid, "You must place a higher bid");

  //   // Update the highest bid
  //   this.highest_bid = { bidder, bid }; // Save the new bid

  //   // Transfer tokens back to the last bidder
  //   return NearPromise.new(lastBidder).transfer(lastBid);
  // }

  @view({})
  get_highest_bid(): Bid {
    return this.highest_bid;
  }

  @view({})
  get_auction_end_time(): BigInt {
    return this.auction_end_time;
  }

  @call({})
  claim() {
    // assert(near.predecessorAccountId() == this.auctioneer, "Only auctioneer can end the auction");
    assert(this.auction_end_time <= near.blockTimestamp(), "Auction has not ended yet");
    assert(!this.auction_was_claimed, "Auction has been claimed");

    this.auction_was_claimed = true;
    return NearPromise.new(this.auctioneer).transfer(this.highest_bid.bid);
  }

  @call({})
  ft_on_transfer({ sender_id, amount, msg }: { sender_id: AccountId, amount: bigint, msg: String }) {

    const previous = { ...this.highest_bid };

    assert(this.auction_end_time > near.blockTimestamp(), "Auction has ended");
    assert(near.predecessorAccountId() == this.ft_contract, "The token is not supported");
    assert(amount >= previous.bid, "You must place a higher bid");

    this.highest_bid = {
      bidder: sender_id,
      bid: amount,
    };

    if (previous.bid > 0) {
      near.log("inside bid");
      // this.ft_transfer(this.highest_bid.bidder, this.highest_bid.bid)
      return NearPromise.new(this.ft_contract)
        .functionCall("ft_transfer", JSON.stringify({ receiver_id: previous.bidder, amount: previous.bid }), BigInt(1), BigInt("30000000000000"))
        .then(
          NearPromise.new(near.currentAccountId())
            .functionCall("ft_transfer_callback", JSON.stringify({}), NO_DEPOSIT, BigInt("30000000000000"))
        )
        .asReturn()
    } else {
      return BigInt(0);
    }
  }

  @call({ privateFunction: true })
  ft_transfer_callback({ }): BigInt {
    return BigInt(0);
  }
}

function promiseResult(i: PromiseIndex): { result: string; success: boolean } {
  near.log("promiseResult");
  let result, success;

  try {
    result = near.promiseResult(i);
    success = true;
  } catch {
    result = undefined;
    success = false;
  }

  return { result, success };
}

File: contract-py/01-basic-auction/contract.py
from near_sdk_py import Contract, view, call, init, Promise


class AuctionContract(Contract):
    """
    A basic auction contract that allows users to place bids.
    """
    
    @init
    def init(self, end_time: int, auctioneer: str):
        """
        Initialize the auction with an end time and auctioneer.
        """
        self.storage["highest_bidder"] = self.current_account_id
        self.storage["highest_bid"] = 1
        self.storage["auction_end_time"] = end_time
        self.storage["auctioneer"] = auctioneer
        self.storage["claimed"] = False
    
    @call
    def bid(self):
        """
        Place a bid in the auction. Must be higher than the current highest bid.
        """
        # Assert the auction is still ongoing
        assert self.block_timestamp < self.storage["auction_end_time"], "Auction has ended"
        
        # Current bid
        bid = self.attached_deposit
        bidder = self.predecessor_account_id
        
        # Last bid
        last_bidder = self.storage["highest_bidder"]
        last_bid = self.storage["highest_bid"]
        
        # Check if the deposit is higher than the current bid
        assert bid > last_bid, "You must place a higher bid"
        
        # Update the highest bid
        self.storage["highest_bidder"] = bidder
        self.storage["highest_bid"] = bid
        
        # Transfer tokens back to the last bidder
        return Promise.create_batch(last_bidder).transfer(last_bid)
    
    @call
    def claim(self):
        """
        Claim the auction proceeds after the auction has ended.
        """
        assert self.block_timestamp > self.storage["auction_end_time"], "Auction has not ended yet"
        assert not self.storage["claimed"], "Auction has already been claimed"
        
        self.storage["claimed"] = True
        
        # Transfer tokens to the auctioneer
        auctioneer = self.storage["auctioneer"]
        highest_bid = self.storage["highest_bid"]
        return Promise.create_batch(auctioneer).transfer(highest_bid)
    
    @view
    def get_highest_bid(self):
        """
        Get the current highest bid.
        """
        return {
            "bidder": self.storage.get("highest_bidder", self.current_account_id),
            "bid": self.storage.get("highest_bid", 1)
        }
    
    @view
    def get_auction_end_time(self):
        """
        Get the auction end time.
        """
        return self.storage.get("auction_end_time", 0)
    
    @view
    def get_auctioneer(self):
        """
        Get the auctioneer account.
        """
        return self.storage.get("auctioneer", "")
    
    @view
    def get_claimed(self):
        """
        Check if the auction has been claimed.
        """
        return self.storage.get("claimed", False)
        